### PR TITLE
1055: District, Provincial and National Level Dashboard: ICT Facilities bar graph

### PR DESCRIPTION
**Pull request by chris-bes**
### Changes:

- Added composeDataByDataPoint builder.
- Added ICT Facilities bar graph

### CODE REVIEW

File: packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * 
+ * Will be converted and returned:
+ * { 
+ *    data: 
+ *        [
+ *          { name: dataPoint1, builder1: 'ABC', builder2: 'GHI' }, 
+ *          { name: dataPoint2, builder1: 'DEF', builder2: 'JKL' }, 
+ *        ]
+ *    }
+ */
```
Comment (chris-bes) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:46]: This is just a reverted version of `composeData` dataBuilder. For this report, we have to return the data like:
```
[
   {
		Pre-School: 0.444,
		Primary: 0.222,
		Secondary: 0.555,
		name: 'Electricity'
   },
   {
		Pre-School: 0.333,
		Primary: 0.444,
		Secondary: 0.555,
		name: 'Functioning smart TV'
	},
]
```

However, it is quite difficult to achieve calculating the percentages for each type of school (stored in attributes column). I guess it is possible, but I can't think of any proper way than adding more config to `percentagesOfValueCounts` data builder which just makes the data builder more complex.

Hence, I went with creating another version of the `composeData` dataBuilder which group the data by the dataPoint (see example documentation in the code)

Comment (chris-bes) [reply packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:46]: Not sure if this is a good solution but happy for any other suggestions!

File: packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ *            dataBuilder: <some databuilder>,
+ *            dataBuilderConfig: <some config>,
+ *        }
+ *    }
+ * }
+ * 
+ * Data returned from fetchComposedData():
+ * {
+ *     builder1: { 
+ *        data: [{name: dataPoint1: value: '111'}, {name: dataPoint2: value: '222'}}
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:31]:
```suggestion
 *        data: [{name: dataPoint1: value: '111'}, {name: dataPoint2: value: '222'}]
```

File: packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ *     },
+ *     builder2: {
+ *        data: [{name: dataPoint1: value: '333'}, {name: dataPoint2: value: '444'}},
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:34]:
```suggestion
 *        data: [{name: dataPoint1: value: '333'}, {name: dataPoint2: value: '444'}],
```

File: packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * { 
+ *    data: 
+ *        [
+ *          { name: dataPoint1, builder1: '111', builder2: '333' }, 
+ *          { name: dataPoint2, builder1: '222', builder2: '444' }, 
+ *        ]
+ *    }
+ */
+
+export const composeDataByDataPoint = async (config, aggregator, dhisApi) => {
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:48]: To keep this consistent with other names, may I suggest `composeDataByDataClass`? Although that is already defined as a helper, so happy to keep it as is, or see if you can somehow replace `dataPoint` with `dataClass`.

File: packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  const responses = await fetchComposedData(config, aggregator, dhisApi);
+  const dataObject = {};
+
+  Object.entries(responses).forEach(([responseName, { data: responseData }]) => {
+    if (!responseData || responseData.length === 0) return;
+    responseData.forEach(responseObject => {
+      const { name: dataPointName, value } = responseObject;
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:55]:
```suggestion
      const { name: dataClassName, value } = responseObject;
```
?

File: packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+      if (!dataObject[dataPointName]) {
+        dataObject[dataPointName] = {};
+        dataObject[dataPointName].name = dataPointName;
+      }
+
+      if (responseObject[`${dataPointName}_metadata`]) {
+        dataObject[dataPointName][`${responseName}_metadata`] = responseObject[`${dataPointName}_metadata`];
+      }
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:64]: Nice feature 🎉

**Review by Alastair-L — APPROVED** [review]
Nice work! I agree that creating a new dataBuilder is the cleanest way to do this. I love that it works at multiple levels out of the box 🎉

File: packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js (hunk 4 above)
Comment (chris-bes) [reply packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:48]: I agree with you, is `composeDataPerDataClass` okay?

Comment (Alastair-L) [reply packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataByDataPoint.js:48]: Yep, that sounds good (although heaven help anyone trying to tell the difference between that and `composeDataByDataClass` 😂. I think it's the best that you could do though with the existing code).

File: packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataPerDataClass.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+ * { 
+ *    data: 
+ *        [
+ *          { name: dataClass1, builder1: '111', builder2: '333' }, 
+ *          { name: dataClass2, builder1: '222', builder2: '444' }, 
+ *        ]
+ * }
+ */
+
+export const composeDataPerDataClass = async (config, aggregator, dhisApi) => {
```
Comment (chris-bes) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataPerDataClass.js:48]: @edmofro Do you care about functional data builders (which doesn't extend from DataBuilder class)?

Comment (edmofro) [reply packages/web-config-server/src/apiV1/dataBuilders/generic/compose/composeDataPerDataClass.js:48]: Sometimes! In this case, there will be no conflict, but if the functional data builder either:
- Use a model through direct import, or 
- Called another functional data builder that relied on models, and didn't pass it in (in this case `fetchComposedData` will receive the models correctly through `config`)

**Pull request merged**